tianshou/utils/net/continuous.py

```diff
@@ -471,3 +472,41 @@ def decode(
         # decode z with state!
         return self.max_action * \
             torch.tanh(self.decoder(torch.cat([state, latent_z], -1)))
+
+
+class GAILDiscriminator(nn.Module):
```

Comment:
Do we really need it? `Critic(Net(...))`

Reply:
I think so because the discriminator needs to take a (state, action)-pair and output a single number. I could use a MLP but then needs to call it as `self.disc(torch.cat([obs, act], dim=1))` instead of `self.disc(obs, act)`.

In AIRL, the discriminator net will be even more complicated, like `self.disc(obs, next_obs, dones, log_pis)`. Then its almost inevitable to add a new class for it.

Reply:
> In AIRL, the discriminator net will be even more complicated, like self.disc(obs, next_obs, dones, log_pis). Then its almost inevitable to add a new class for it.

Here are my first thoughts: 
1. Is it possible to design the APIs here for AIRL?
2. Is it possible to inherit from Critic?

Reply:
You have a point. I refactored the code and now it didn't need a special discriminator class and it would be easy to extend to AIRL by overriding `def disc(self, batch: Batch)` method since all the information (obs, next_obs, dones, log_pis) can be passed through a `batch`.